langchain/retrievers/merger_retriever.py

```diff
@@ -0,0 +1,165 @@
+from langchain.schema import BaseRetriever
+from langchain.schema import BaseDocumentTransformer, Document
+from typing import List
+import asyncio
+
+
+class MergerRetriever(BaseRetriever):
+    """
+    This class merges the results of multiple retrievers and then refines the results using a list of transformers.
+
+    Args:
+        retrievers: A list of retrievers to merge.
+        transformers: A list of transformers to refine the results.
+    """
+
+    def __init__(
+        self,
+        retrievers: List[BaseRetriever],
+        transformers: List[BaseDocumentTransformer],
+    ):
+        """
+        Initialize the MergerRetriever class.
+
+        Args:
+            retrievers: A list of retrievers to merge.
+            transformers: A list of transformers to refine the results.
+        """
+
+        self.retrievers = retrievers
+        self.transformers = transformers
+
+    def get_relevant_documents(self, query: str) -> List[str]:
+        """
+        Get the relevant documents for a given query.
+
+        Args:
+            query: The query to search for.
+
+        Returns:
+            A list of relevant documents.
+        """
+
+        # Merge the results of the retrievers.
+        merged_documents = self.merge_documents(query)
+
+        # Refine the results using the transformers.
+        refined_documents = self.refine_documents(merged_documents)
+
+        return refined_documents
+
+    async def aget_relevant_documents(self, query: str) -> List[str]:
+        """
+        Asynchronously get the relevant documents for a given query.
+
+        Args:
+            query: The query to search for.
+
+        Returns:
+            A list of relevant documents.
+        """
+
+        # Merge the results of the retrievers.
+        merged_documents = await self.amerge_documents(query)
+
+        # Refine the results using the transformers.
+        refined_documents = await self.arefine_documents(merged_documents)
+
+        return refined_documents
+
+def merge_documents(self, query: str) -> List[str]:
+        """
+        Merge the results of the retrievers.
+
+        Args:
+            query: The query to search for.
+
+        Returns:
+            A list of merged documents.
+        """
+
+        # Get the results of all retrievers.
```

Comment:
not sure how much better this is, but from https://docs.python.org/3/library/itertools.html#itertools-recipes:
```python
def roundrobin(*iterables):
    "roundrobin('ABC', 'D', 'EF') --> A D E B F C"
    # Recipe credited to George Sakkis
    num_active = len(iterables)
    nexts = cycle(iter(it).__next__ for it in iterables)
    while num_active:
        try:
            for next in nexts:
                yield next()
        except StopIteration:
            # Remove the iterator we just exhausted from the cycle.
            num_active -= 1
            nexts = cycle(islice(nexts, num_active))
```

Reply:
That's fancy logic right there! I never thought this was a roundrobin use case but  if kinda fits. 
Downside: It's a little harder to read.
I was thinking that we could expose a couple of diff merging strategies with a configurable parameter.